Mathlib/Algebra/MonoidAlgebra/NoZeroDivisors.lean

```diff
@@ -46,119 +50,60 @@ namespace AddMonoidAlgebra
-    obtain ⟨a, ha, H⟩ :=
-      Left.exists_add_of_mem_support_single_mul fmin
-        ((single fmin 1 * g : AddMonoidAlgebra R A).support.min'
-          (by rw [support_single_mul] <;> simp [support_nonempty_iff.mpr fg.2]))
-        (Finset.min'_mem _ _)
-    refine' ⟨fmin + a, mem_support_iff.mpr _⟩
-    rw [mul_apply_add_eq_mul_of_forall_ne _]
-    · refine' mul_ne_zero _ _
-      exacts [mem_support_iff.mp (Finset.min'_mem _ _), mem_support_iff.mp ha]
-    · rw [H]
-      rintro b c bf cg (hb | hc) <;> refine' ne_of_gt _
-      · refine' lt_of_le_of_lt (_ : _ ≤ fmin + c) _
-        · apply Finset.min'_le
-          rw [support_single_mul _ _ (fun y => by rw [one_mul] : ∀ y : R, 1 * y = 0 ↔ _)]
-          simp only [cg, Finset.mem_map, addLeftEmbedding_apply, add_right_inj, exists_prop,
-            exists_eq_right]
-        · refine' add_lt_add_right _ _
-          exact Finset.min'_lt_of_mem_erase_min' _ _ (Finset.mem_erase.mpr ⟨hb, bf⟩)
-      · refine' lt_of_lt_of_le (_ : _ < fmin + c) _
-        · apply Finset.min'_lt_of_mem_erase_min'
-          rw [← H]
-          apply Finset.mem_erase_of_ne_of_mem
-          · simpa only [Ne.def, add_right_inj]
-          · rw [support_single_mul _ _ (fun y => by rw [one_mul] : ∀ y : R, 1 * y = 0 ↔ _)]
-            simpa only [Finset.mem_map, addLeftEmbedding_apply, add_right_inj, exists_prop,
-              exists_eq_right]
-        · haveI : CovariantClass A A (Function.swap (· + ·)) (· ≤ ·) :=
-            Add.to_covariantClass_right A
-          exact add_le_add_right (Finset.min'_le _ _ bf) _⟩
-#align add_monoid_algebra.no_zero_divisors.of_right_ordered AddMonoidAlgebra.NoZeroDivisors.of_right_ordered
-
-end LeftOrRightOrderability
+  simp_rw [mul_apply, sum, ← Finset.sum_product']
+  refine' (Finset.sum_eq_single (a0, b0) _ _).trans (if_pos rfl) <;> simp_rw [Finset.mem_product]
+  · refine fun ab hab hne => if_neg (fun he => hne <| Prod.ext ?_ ?_)
+    exacts [(h hab.1 hab.2 he).1, (h hab.1 hab.2 he).2]
+  · refine fun hnmem => ite_eq_right_iff.mpr (fun _ => ?_)
+    rcases not_and_or.mp hnmem with af | bg
+    · rw [not_mem_support_iff.mp af, zero_mul]
+    · rw [not_mem_support_iff.mp bg, mul_zero]
+
+instance {A : Type*} [NoZeroDivisors R] [Add A] [UniqueSums A] :
+    NoZeroDivisors (AddMonoidAlgebra R A) where
+  eq_zero_or_eq_zero_of_mul_eq_zero := fun {a b} ab => by
+    contrapose! ab
+    obtain ⟨da, a0, db, b0, h⟩ := UniqueSums.uniqueAdd_of_nonempty
+      (support_nonempty_iff.mpr ab.1) (support_nonempty_iff.mpr ab.2)
+    refine support_nonempty_iff.mp ⟨da + db, ?_⟩
+    rw [mem_support_iff] at a0 b0 ⊢
+    exact mul_apply_add_eq_mul_of_uniqueAdd h ▸ mul_ne_zero a0 b0
+
+/-- The proof goes via the equivalence `R ≃ₗ[ℚ] (Basis.ofVectorSpaceIndex ℚ R) →₀ ℚ`,
+i.e. choosing a basis.
+Once we have a basis, we use the Lexicographic order on the coordinates and all the instances
+that `ℚ` already has.
+-/
+instance {R : Type*} [AddCommGroup R] [Module ℚ R] : UniqueSums R :=
```

Comment:
By the way, `Module ℚ R` could be replaced by torsionfree.  I chose this assumption as it was a cheap way out and it still worked for Lindemann-Weierstrass.

In the longer term, proving that torsion-free abelian groups have `UniqueSums` could be nice.

Reply:
We just need the fact that [torsion-free abelian groups embed into a rational vector space](https://math.stackexchange.com/questions/425431/embedding-torsion-free-abelian-groups-into-mathbb-qn). Torsion-free abelian cancellative monoids embeds into such groups, so it even extends to that case.

Reply:
To show this with what we currently have in mathlib:
1. for a commutative cancellative monoid, we know it embeds into its AddLocalization at \top ([AddLocalization.mk_left_injective](https://leanprover-community.github.io/mathlib4_docs/Mathlib/GroupTheory/MonoidLocalization.html#AddLocalization.mk_left_injective)) but we don't have that the AddLocalization is a group. We also need to show [AddMonoid.IsTorsionFree](https://leanprover-community.github.io/mathlib4_docs/Mathlib/GroupTheory/Torsion.html#AddMonoid.IsTorsionFree) is preserved by AddLocalization.
2. Show that `AddMonoid.IsTorsionFree A` is equivalent to `NoSmulZeroDivisors ℕ A` (or `NoSmulZeroDivisors ℤ A` for AddGroup A).
3. Show that `[LocalizedModule.mkLinearMap](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Module/LocalizedModule.html#LocalizedModule.mkLinearMap) (NonZeroDivisors R) M` is injective when `NoSmulZeroDivisors R M` is satisfied: this should be easy from [LocalizedModule.mk_eq](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Module/LocalizedModule.html#LocalizedModule.mk_eq). `LocalizedModule ℤ⁰ M` is a module over `Localization ℤ⁰` (= `FractionRing ℤ`) by [LocalizedModule.isModule](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Module/LocalizedModule.html#LocalizedModule.isModule), so `LocalizedModule ℤ⁰ M` is a vector space over `FractionRing ℤ` and therefore isomorphic to some `ι →₀ FractionRing ℤ`, which is isomorphic (as AddGroup) to `ι →₀ ℚ`, and we can then apply the instance here.